src/axom/spin/internal/linear_bvh/build_radix_tree.hpp

```diff
@@ -487,47 +480,60 @@ void propagate_aabbs(RadixTree<FloatType, NDIMS>& data, int allocatorID)
         int32 lchild = lchildren_ptr[current_node];
         int32 rchild = rchildren_ptr[current_node];
 
-        // gather the aabbs
-        primal::BoundingBox<FloatType, NDIMS> aabb;
-        if(lchild >= inner_size)
+        int other_child = (lchild == last_node) ? rchild : lchild;
+
+        if(other_child >= inner_size)
         {
-          aabb.addBox(leaf_aabb_ptr[lchild - inner_size]);
+          aabb.addBox(leaf_aabb_ptr[other_child - inner_size]);
         }
         else
         {
-          aabb.addBox(inner_aabb_ptr[lchild]);
+          PointType other_aabb_min, other_aabb_max;
+          for(int idim = 0; idim < NDIMS; idim++)
+          {
+            // NOTE: These atomicAdd(..., 0)  operations are to ensure that
+            // we get the latest value from L2$. Might not be necessary, since
+            // the atomic store below should bypass L1$ entirely, but just in
+            // case...
+            other_aabb_min[idim] =
+              RAJA::atomicAdd<atomic_policy>(&min_range[other_child][idim],
+                                             FloatType {0});
+            other_aabb_max[idim] =
+              RAJA::atomicAdd<atomic_policy>(&max_range[other_child][idim],
+                                             FloatType {0});
+          }
+          aabb.addPoint(other_aabb_min);
+          aabb.addPoint(other_aabb_max);
         }
 
-        if(rchild >= inner_size)
-        {
-          aabb.addBox(leaf_aabb_ptr[rchild - inner_size]);
-        }
-        else
+        for(int idim = 0; idim < NDIMS; idim++)
         {
-          aabb.addBox(inner_aabb_ptr[rchild]);
+          // Store our final AABB value for the current node to global memory
+          // coherently. Atomics are resolved at the L2 cache, so this should
+          // be observable by all threads.
+          RAJA::atomicAdd<atomic_policy>(&min_range[current_node][idim],
+                                         aabb.getMin()[idim]);
```

Comment:
More specifically, should these be `atomicMin()` and `atomicMax()` calls instead of `atomicAdd()`

Reply:
The bounding box calculations for the inner node actually happen in a different place:

https://github.com/LLNL/axom/blob/c677e4507f8a07643e90ef4fc4b2b3178315a7a6/src/axom/spin/internal/linear_bvh/build_radix_tree.hpp#L485-L507

Only one thread handles the computation and writing of a bounding box for a given inner node (the first child to reach an internal node exits), so we don't need `atomicMin()` and `atomicMax()`. Instead, the `atomicAdd()`s are pretty much used as a "load/store directly to L2" operation.

We could have also achieved uncached loads with the use of `volatile FloatType*`, but I think the atomic solution is a little cleaner (and also lets us keep using primal types)

Reply:
Thanks for the explanation Max -- I see it now -- one is a load and the other is a store.

It might be clearer if these were expressed as (inlined) functions/macros to make it more obvious.